### PR TITLE
Add class docstring to NodeMeta

### PR DESCRIPTION
**Pull request by WarrenTheRabbit**
#### Overview
I attempt to add a docstring for `NodeMeta` that contains a helpful summary of what it does and the reason for its existence, as well as a link to a project board that contains more information. The intention was to make the docstring helpful to new and experienced contributors alike.

Note: I did a direct edit on the pytest repo but I am unsure if it came through. If the direct edit and this PR clash, then this PR is the authoritative contribution.

#### Areas of Concern
1) I am unsure whether my summary is accurate or current, so I'd be especially grateful if @RonnyPfannschmidt or others involved with detangling `Node` relationships could review this.

2) I don't have much experience with Sphinx documentation. I tried to use directives but I am unsure if linking to a project board should be handled in a particular way.

#### Housekeeping
- [X] Allow maintainers to push and squash when merging my commits. Please uncheck this if you prefer to squash the commits yourself.

### CODE REVIEW

**Comment by nicoddemus** [comment]
> I don't have much experience with Sphinx documentation. I tried to use directives but I am unsure if linking to a project board should be handled in a particular way.

In this case it does not matter much, because `NodeMeta` is not documented on purpose, we consider it an implementation detail.

**Review by nicoddemus — APPROVED** [review]
LGTM!

Leaving to @RonnyPfannschmidt to do a final review/merge.

**Comment by WarrenTheRabbit** [comment]
> > @WarrenTheRabbit : . . . unsure if linking to a project board should be handled in a particular way.
> 
> @nicoddemus: In this case it does not matter much, because `NodeMeta` is not documented on purpose

That makes sense! Thank you for the context. I'm now trying to figure out how Sphinx is discovering which parts of the codebase to document or which parts to specifically exclude. 

Is excluding NodeMeta from Sphinx's documentation processes currently achieved explicitly, or is it happening only because its docstring is empty?

I'm wondering if I need to handle the exclusion explicitly now that a docstring has been added. I'm also wondering if I should remove the directives because a) they add clutter that (at least for me) is not rendered nicely by my IDE and b) I don't want to imply they are built in the docs if they aren't.

File: src/_pytest/nodes.py [hunk 1]
```diff
@@ -127,6 +127,20 @@ def _imply_path(
 
 
 class NodeMeta(type):
```
Comment (WarrenTheRabbit) [inline src/_pytest/nodes.py:129]: @nicoddemus mentioned elsewhere that _ indicates private but also that a private access modifier is implied when already inside an internal module. 

As an implementation detail, I'm wondering if `NodeMeta` should be so designated: `_NodeMeta`.
Or is it already implied by its parent directory `_pytest`?

Comment (RonnyPfannschmidt) [reply src/_pytest/nodes.py:129]: its already implied by _pytest

**Comment by RonnyPfannschmidt** [comment]
we might want to add a part to the documentation thats intended to help developers of the internals there we could add such references

is there a easy way to partition the reference docs to do this (I haven't done such a split before)

**Review by RonnyPfannschmidt — APPROVED** [review]
preliminary approve with the intent to add the link to a "internals for people working on pytest" part of the documentation

depending on the ease of adding such section we should either merge with a followup issue, or add a bare bones version of the section before

thanks for getting this discussion started

its always great to have someone new, not settled "in the ways of the project" come in and experiment with improvements and/or asking the questions that feel oblivious only when there is experience and bias - because it gives opportunity to turn things that ought to be obvious from "obvious for the core-dev" to "obvious for everyone"

**Comment by nicoddemus** [comment]
> Is excluding NodeMeta from Sphinx's documentation processes currently achieved explicitly, or is it happening only because its docstring is empty?

We explicitly declare which classes are documented, so we just do not declare `NodeMeta`:

https://github.com/pytest-dev/pytest/blob/6c2feb75d2c4bb01aa145f8b85f7fb09fe4133cf/doc/en/reference/reference.rst?plain=1#L796-L800

> preliminary approve with the intent to add the link to a "internals for people working on pytest" part of the documentation

We just need to create a separate page and declare the internal classes there.

**Comment by WarrenTheRabbit** [comment]
> @RonnyPfannschmidt: thanks for getting this discussion started . . . its always great to have someone new . . . because it gives opportunity to turn things that ought to be obvious from "obvious for the core-dev" to "obvious for everyone"

Thank you, @RonnyPfannschmidt, your continued welcome and support is very meaningful to me. Likewise to @nicoddemus and @bluetech. I'm aware that I'm repeating myself by giving thanks again here, as I've already expressed warm sentiments elsewhere, but some things are worth repeating. One can't always be *dry*. Ba-boom.  

While speaking from the heart, I'd like to say I'm full of admiration for everything that has been achieved by you, the current team, all who have been before and those who still try to pop in and out whenever they can. 

My life is hectic, I'm desperately trying to get an entry level job here in Australia and the repository constantly challenges me but I know I'll be sticking it out for the long-term, staying involved and doing what I can to steadily contribute more and more as my capabilities increase. I couldn't be happier to be part of pytest, and without the work that has already been done and the welcome that is always being given, I wouldn't have the opportunity and there would be less satisfaction and happiness in my life. So very sincerely, thank you once again to all.

**Comment by WarrenTheRabbit** [comment]
> @RonnyPfannschmidt: depending on the ease of adding such section we should either merge with a followup issue, or add a bare bones version of the section before

I'm thinking it would be better to merge first because it adds a docstring that may be helpful and causes no harm. Then, when the other section is created, there is the added advantage that there will be something to populate it with immediately. Unless I am missing an important ordering effect?

Also, @nicoddemus shone a light on how Sphinx might be working. It is looking like strategic edits will mostly be all that is required to get a new section up (depending on the design chosen). I'm thinking I'll open a new issue to 

1.  give and discuss design options (an entirely new page or a subsection in the [Reference guide](https://docs.pytest.org/en/latest/reference/index.html#), for example)
2. outline what I think needs to be done to implement the designs

Then either myself or someone else will tackle the PR. What do you think?

**Comment by nicoddemus** [comment]
> So very sincerely, thank you once again to all.

Sure thing, and once again, thanks for the enthusiasm and willingness to contribute. We really appreciate it and are glad to have you as part of the ecosystem! 👍 

> Then either myself or someone else will tackle the PR. What do you think?

Agree with both suggestions: merge this now and create a new issue to tackle the problem of "internal implementation reference". 👍

**Pull request merged**